project/improved_algorithm_interface.py

```diff
@@ -0,0 +1,354 @@
+    print "The average group size is " + \
+    str(group_length/(len(Groups) - empty_group))
+    print "There are " + str(empty_group) + " empty groups"
+
+#***********************
+# Main Program
+# ***********************
+
+# ------------------
+# Initialize Data
+# ------------------
+
+
+def get_students():
+    students = []
+    student_query = session.query(Student).all()
+    preference_query = None
+    count = 1
+    # makes a list of Student objects from the database
+    for student in student_query:
+        preference_query = session.query(Preference).filter_by(student_id=count).all()
+        students.append(Student_class(student.first_name, -1, -1, False, preference_query))
+        count = count + 1
+    return students
+
+
+def sort_students(students):
+    students.sort(key=lambda students: (students.pref_score, students.mutual_score), reverse=True)
+    return students
+
+
+def score_students(students):
+    for primary_student in students:
+        for comparison_student in students:
+            for index in range(max_prefs):
+                if(comparison_student.name == primary_student.prefs[index].name):
+                    comparison_student.pref_score += 1
+                    for secondary_index in range(len(primary_student.prefs)):
+                        if(comparison_student.prefs[secondary_index].name == primary_student.name):
+                            comparison_student.mutual_score += 1
+    return None
+
+
+def convert_pref_student(students):
+    pref = []
+    for primary_student in students:
+        for comparison_student in students:
+            for index in range(4):
+                if(comparison_student.name == primary_student.prefs[index].name):
+                    pref.append(comparison_student)
+        primary_student.prefs = pref
+        pref = []
+
+
+def concatenate_names(students):
+    for student in students:
+        for index in range(len(student.prefs)):
+            student.prefs[index].name = student.prefs[index].first_name1 + student.prefs[index].last_name1
+    return None
+
+
+# Create Groups
+
+Groups = []
+
+
+def setup():
+    for iterate in range(num_of_groups):  # initialize groups
+        Groups.append([])
+
+
+# -------------------------------
+# Assign students to Groups
+# -------------------------------
+def order_groups(students, student):
+    group_order = []
+
+# get info for each group
+    for count in range(num_of_groups):
+        group_order.append(["", "", ""])
+        group_order[count][0] = count
+        group_order[count][1] = len(Groups[count])
+        lowest_pref = pref_in_group(student, Groups[count])
+        group_order[count][2] = lowest_pref.pref_score
+
+# sort group_order by groupSize then by lowest preference score of a preference
+    group_order.sort(
+        key=lambda l: (l[1], l[2])
+    )
+    return group_order
+
+def condition_g(try_to_pull_preference_with_student, student, group_order, group_index):
```

Comment:
The conditions should be in order and directly follow each other. We can talk about clearer method names (we would like someone to be able to read through the code and see what's going on rather than have to read the outside document)